service/matching/version_rule_helpers.go

```diff
@@ -491,6 +501,51 @@ func dfs(curr string, visited, inStack map[string]bool, nodes map[string][]strin
 	return false
 }
 
+// getUpstreamBuildIds returns a list of build ids that point to the given buildId in the graph of redirect rules.
+// It considers all rules in the input list, so the caller should provide a filtered list as needed.
+// The result will contain no duplicates.
+func getUpstreamBuildIds(buildId string, redirectRules []*persistencespb.RedirectRule) []string {
+	return getUpstreamHelper(buildId, redirectRules, nil)
+}
+
+func getUpstreamHelper(
+	buildId string,
+	redirectRules []*persistencespb.RedirectRule,
+	visited []string) []string {
+	var upstream []string
+	visited = append(visited, buildId)
+	directSources := getSourcesForTarget(buildId, redirectRules)
+
+	for _, src := range directSources {
+		if !slices.Contains(visited, src) {
+			upstream = append(upstream, src)
+			upstream = append(upstream, getUpstreamHelper(src, redirectRules, visited)...)
+		}
+	}
+
+	// dedupe
+	upstreamUnique := make(map[string]bool)
+	for _, bid := range upstream {
+		upstreamUnique[bid] = true
+	}
+	upstream = make([]string, 0)
```

Comment:
You know how large this will be so can pre-size the array
```suggestion
    upstream = make([]string, 0, len(upstreamUnique))
```

Reply:
same as above, I considered doing this but since we know that the total length is always bounded, I didn't think it was worth the extra two lines of code to create and update an index counter

Reply:
done